File: echo-web/src/main/java/com/netflix/spinnaker/echo/config/Front50Config.java
/*
 * Copyright 2018 Google, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.netflix.spinnaker.echo.config;

import com.netflix.spinnaker.config.OkHttp3ClientConfiguration;
import com.netflix.spinnaker.echo.services.Front50Service;
import com.netflix.spinnaker.kork.retrofit.ErrorHandlingExecutorCallAdapterFactory;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import retrofit2.Retrofit;
import retrofit2.converter.jackson.JacksonConverterFactory;

@Configuration
@Slf4j
public class Front50Config {

  @Bean
  public Front50Service front50Service(
      @Value("${front50.base-url}") String front50BaseUrl,
      OkHttp3ClientConfiguration okHttp3ClientConfiguration) {
    log.info("front50 service loaded");

    return new Retrofit.Builder()
        .baseUrl(front50BaseUrl)
        .client(okHttp3ClientConfiguration.createForRetrofit2().build())
        .addCallAdapterFactory(ErrorHandlingExecutorCallAdapterFactory.getInstance())
        .addConverterFactory(JacksonConverterFactory.create())
        .build()
        .create(Front50Service.class);
  }
}

File: keel-front50/src/main/kotlin/com/netflix/spinnaker/config/Front50Config.kt
package com.netflix.spinnaker.config

import com.fasterxml.jackson.databind.ObjectMapper
import com.netflix.spinnaker.keel.front50.Front50Service
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.OkHttpClient
import org.springframework.beans.factory.BeanCreationException
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import retrofit2.Retrofit
import retrofit2.converter.jackson.JacksonConverterFactory

@Configuration
@ConditionalOnProperty("front50.enabled")
class Front50Config {
  @Bean
  fun front50Endpoint(@Value("\${front50.base-url}") front50BaseUrl: String): HttpUrl =
    front50BaseUrl.toHttpUrlOrNull()
      ?: throw BeanCreationException("Invalid URL: $front50BaseUrl")

  @Bean
  fun front50Service(
    front50Endpoint: HttpUrl,
    objectMapper: ObjectMapper,
    retrofitClient: OkHttpClient
  ): Front50Service =
    Retrofit.Builder()
      .addConverterFactory(JacksonConverterFactory.create(objectMapper))
      .baseUrl(front50Endpoint)
      .client(retrofitClient)
      .build()
      .create(Front50Service::class.java)
}
